content_scripts/marks.coffee

```diff
@@ -1,45 +1,75 @@
-        scrollX: window.scrollX,
-        scrollY: window.scrollY
-      HUD.showForDuration "Created local mark '#{keyChar}'", 1000
-
-    @remove()
-
-    false
-
-root.activateGotoMode = ->
-  handlerStack.push keydown: (e) ->
-    keyChar = KeyboardUtils.getKeyChar(event)
-    return unless keyChar isnt ""
-
-    if /[A-Z]/.test keyChar
-      chrome.runtime.sendMessage
-        handler: 'gotoMark'
-        markName: keyChar
-    else if /[a-z]/.test keyChar
-      [baseLocation, sep, hash] = window.location.href.split '#'
-      markString = localStorage["vimiumMark|#{baseLocation}|#{keyChar}"]
-      if markString?
-        mark = JSON.parse markString
-        window.scrollTo mark.scrollX, mark.scrollY
-        HUD.showForDuration "Jumped to local mark '#{keyChar}'", 1000
-
-    @remove()
-
-    false
+
+Marks =
+  mode: null
+  previousPosition: null
+
+  exit: (continuation = null) ->
+    @mode?.exit()
+    @mode = null
+    continuation?()
+
+  # This returns the key which is used for storing mark locations in localStorage.
+  getLocationKey: (keyChar) ->
+    "vimiumMark|#{window.location.href.split('#')[0]}|#{keyChar}"
+
+  getMarkString: ->
+    JSON.stringify scrollX: window.scrollX, scrollY: window.scrollY
+
+  setPreviousPosition: ->
+    @previousPosition = @getMarkString()
+
+  showMessage: (message, keyChar) ->
+    HUD.showForDuration "#{message} \"#{keyChar}\".", 1000
+
+  activateCreateMode: ->
+    @mode = new Mode
+      name: "create-mark"
+      indicator: "Create mark..."
+      suppressAllKeyboardEvents: true
+      keypress: (event) =>
+        keyChar = String.fromCharCode event.charCode
+        # If <Shift> is depressed, then it's a global mark, otherwise it's a local mark.  This is consistent
+        # vim's [A-Z] for global marks, [a-z] for local marks.  However, it also admits other non-Latin
+        # characters.
+        @exit =>
+          if event.shiftKey
+              chrome.runtime.sendMessage
+                handler: 'createMark'
+                markName: keyChar
+                scrollX: window.scrollX
+                scrollY: window.scrollY
+              , => @showMessage "Created global mark", keyChar
+          else
+              localStorage[@getLocationKey keyChar] = @getMarkString()
+              @showMessage "Created local mark", keyChar
+
+  activateGotoMode: (registryEntry) ->
+    # We pick off the last character of the key sequence used to launch this command. Usually this is just "`".
+    # We then use that character, so together usually the sequence "``", to jump back to the previous
+    # position.  The "previous position" is recorded below, and is registered via @setPreviousPosition()
+    # elsewhere for various other jump-like commands.
+    previousPositionKey = registryEntry.key[registryEntry.key.length-1..]
+    @mode = new Mode
+      name: "goto-mark"
+      indicator: "Go to mark..."
+      suppressAllKeyboardEvents: true
+      keypress: (event) =>
```

Comment:
The difference here is that there's no config a user can tweak to make multiple layouts consistent, whereas normal mode can be made consistent just by adding the right mappings. If we do something weird here, then we cause pain that users can't mitigate.

Maybe there can be a compromise. Looking at `keypress` it seems as though we should be able to extract `keyCode` and `charCode`. I think it deals with both of our concerns if we
* associate the character (derived from `charCode`) *and* the key (`keyCode`) with each mark
* try to go to the mark for the character when asked
* if that fails, try to go to the mark for the `keyCode`
* if that also fails, show the error message.

This way, everybody should be able to access the mark they meant to from any keyboard layout. Is this plausible?